### PR TITLE
WIP - Vioscreen Report PDF

### PR DESCRIPTION
**Pull request by dhakim87**
Adds support for talking to vioscreen using RPC through celery.  Have to worry about both the RPC channel only allowing jsonifiable data, handling async correctly, and dealing with content types and pdf embedding, but it all works as a prototype.  Note that the initial vioscreen login handshake takes about 35 seconds (as presumably does any token refresh), but once logged in the pdf generation time is near instant.  Must check if vioscreen meters our pdf generation and we're forced to cache results.

Uncovered a potential issue where survey ids in developer builds are referring to both valid vioscreen data and primary survey data, which is not supported by our model classes/repos.

### CODE REVIEW

File: microsetta_private_api/util/vioscreen.py [hunk 1]
```diff
@@ -79,3 +86,214 @@ def decode_key(encoded):
...
+        method = self.session.get
+    elif method == "POST":
+        method = self.session.post
+    else:
+        raise Exception("Unknown Method")
+
+    url = urljoin(self.baseurl, url)
+
+    req = None
+    for try_twice in range(2):
```
Comment (gwarmstrong) [inline microsetta_private_api/util/vioscreen.py:162]:
```suggestion
    for _ in range(2):
```

File: microsetta_private_api/util/vioscreen.py [hunk 2]
```diff
@@ -79,3 +86,214 @@ def decode_key(encoded):
...
+        req = method(url, headers=self.headers, **kwargs)
+        if req.status_code != 200:
+            data = req.json()
+            code = data.get('Code')
+
+            if code == 1016:
```
Comment (gwarmstrong) [inline microsetta_private_api/util/vioscreen.py:168]: Where do these codes come from?

File: microsetta_private_api/util/vioscreen.py (hunk 1 above)
Comment (gwarmstrong) [reply microsetta_private_api/util/vioscreen.py:162]: The logic of why you are doing this is a loop is a little confusing.

What about moving things to a function that returns `(data, needs_retry)`

Then something like:
```
data, needs_retry = get_pdf(...)
if needs_retry:
    self.update_headers()
    data, needs_retry = get_pdf(...)
```

File: microsetta_private_api/util/vioscreen.py [hunk 3]
```diff
@@ -79,3 +86,214 @@ def decode_key(encoded):
...
+    def _get_session_data(self, id_, name):
+        url = 'sessions/%s/%s' % (id_, name)
+        result = self.get(url)
+        if 'error' in result:
+            return {}
+        else:
+            return result
+
+    def foodcomponents(self, id_):
+        return self._get_session_data(id_, 'foodcomponents').get('data')
```
Comment (gwarmstrong) [inline microsetta_private_api/util/vioscreen.py:232]: Feels weird that you cannot get all this info with a single RPC. Is there not a way to query all of this info at once?

**Review by gwarmstrong — COMMENTED** [review]
Looks good. A few comments.

ALSO: looks like you have failing tests, can you address those?

File: microsetta_private_api/api/_survey.py [hunk 4]
```diff
@@ -226,3 +228,33 @@ def _submit_vioscreen_status(account_id, source_id, info_str):
...
+        status = vioscreen_repo.get_vioscreen_status(account_id,
+                                                     source_id,
+                                                     survey_id)
+        if status != 3:
+            # Oops, we don't have results available for this one
+            raise NotFound("No such survey recorded")
+
+        vio = VioscreenAdminAPI()
+        sessions = vio.sessions(survey_id)
+        # Looks like vioscreen supports multiple sessions per user, do we care?
```
[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/api/_survey.py:249]: I don't believe we're relying on multiple sessions

File: microsetta_private_api/api/_survey.py [hunk 5]
```diff
@@ -226,3 +228,33 @@ def _submit_vioscreen_status(account_id, source_id, info_str):
...
+    with Transaction() as t:
+        vioscreen_repo = VioscreenRepo(t)
+
+        # Vioscreen username is our survey_id
+        status = vioscreen_repo.get_vioscreen_status(account_id,
+                                                     source_id,
+                                                     survey_id)
+        if status != 3:
+            # Oops, we don't have results available for this one
+            raise NotFound("No such survey recorded")
```
[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_private_api/api/_survey.py:245]: The NotFound will bubble up to Flask's error handlers and throw a 404 for the requested route.  It should be clear from the context of the url what the account id, source id and survey id are if a user hits the page (and will also be ignored at the moment by the microsetta-interface project since no user actually hits these routes directly).  Unless/until we move to a system with application layer error codes that can be parsed and handled in code, I don't see a reason to extend the 404 beyond a string I can grep for if it appears in an error report.  

Would punt for now.

File: microsetta_private_api/api/microsetta_private_api.yaml [hunk 6]
```diff
@@ -681,6 +681,32 @@ paths:
...
+      summary: Generate vioscreen top food report
+      description:  Generate vioscreen top food report
+      parameters:
+        - $ref: '#/components/parameters/account_id'
+        - $ref: '#/components/parameters/source_id'
+        - $ref: '#/components/parameters/survey_id'
+        - $ref: '#/components/parameters/language_tag'
+      responses:
+        '200':
+          description: Top Food Report pdf
```
[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_private_api/api/microsetta_private_api.yaml:702]: Definitely.  I spent 20 minutes and couldn't find the magic sauce to make yaml stop complaining.  I'll give this a shot.

[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_private_api/api/microsetta_private_api.yaml:702]: Yep, nope.  Connexion explodes trying to verify the pdf response with these settings on output.  I'm sure there is a proper schema that can be assigned, but I have no idea how to do it.

File: microsetta_private_api/api/_survey.py [hunk 7]
```diff
@@ -226,3 +228,33 @@ def _submit_vioscreen_status(account_id, source_id, info_str):
...
+
+        vio = VioscreenAdminAPI()
+        sessions = vio.sessions(survey_id)
+        # Looks like vioscreen supports multiple sessions per user, do we care?
+        session_id = sessions[0]['sessionId']
+        report = vio.top_food_report(session_id)
+
+        response = make_response(report)
+        response.headers.set("Content-Type", "application/pdf")
+        # TODO: Do we want it to download a file or be embedded in the html?
```
Comment (wasade) [inline microsetta_private_api/api/_survey.py:255]: Not sure of the pros/cons here

File: microsetta_private_api/repo/survey_answers_repo.py [hunk 8]
```diff
@@ -44,18 +44,23 @@ def find_survey_template_id(self, survey_answers_id):
...
-                if status is not None:
-                    return SurveyTemplateRepo.VIOSCREEN_ID
-                else:
-                    return None
-                    # TODO: Maybe this should throw an exception, but doing so
-                    #  locks the end user out of the minimal implementation
-                    #  if they submit an empty survey response.
-                    # raise RepoException("No answers in survey: %s" %
-                    #                     survey_answers_id)
+            # TODO:  We probably can't merge the PR with this change, need
```
Comment (wasade) [inline microsetta_private_api/repo/survey_answers_repo.py:47]: should we put in a patch in a separate PR to resolve this?

File: microsetta_private_api/repo/survey_answers_repo.py [hunk 9]
```diff
@@ -44,18 +44,23 @@ def find_survey_template_id(self, survey_answers_id):
...
+            # if len(rows) == 0:
+            vioscreen_repo = VioscreenRepo(self._transaction)
+            status = vioscreen_repo._get_vioscreen_status(survey_answers_id)
+            if status is not None:
+                return SurveyTemplateRepo.VIOSCREEN_ID
+            else:
+                return None
+                # TODO: Maybe this should throw an exception, but doing so
+                #  locks the end user out of the minimal implementation
+                #  if they submit an empty survey response.
```
Comment (wasade) [inline microsetta_private_api/repo/survey_answers_repo.py:61]: Do we have a notion of an empty survey response here?

File: microsetta_private_api/util/vioscreen.py [hunk 10]
```diff
@@ -79,3 +86,221 @@ def decode_key(encoded):
...
+class VioscreenAdminAPITask(Task):
+    def __init__(self):
+        print("Initializing Vioscreen Communications Task")
+        self.baseurl = urljoin('https://api.viocare.com',
+                               SERVER_CONFIG["vioscreen_regcode"]) + '/'
+        self.user = SERVER_CONFIG["vioscreen_admin_username"]
+        self.password = SERVER_CONFIG["vioscreen_admin_password"]
+        self.session = requests.Session()
+        self.headers = None
+        # TODO:  Rage.  The flask side instantiates this task upon doing RPC
```
Comment (wasade) [inline microsetta_private_api/util/vioscreen.py:112]: I *think* it's getting instantiated on the `@celery.task()` decorator when this file is imported

Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:112]: One alternative that would (a) avoid the instance getting triggered by flask on module import and (b) be process specific is below:

```python
VioscreenAdminAPITask_singleton = None

@celery.task(
    ignore_result=False,  # We need the results back
    default_retry_delay=5  # Default policy: wait 5 seconds on failure
)
def make_vioscreen_request(method, url, **kwargs):
    global VioscreenAdminAPITask_singleton
    if VioscreenAdminAPITask_singleton is None:
        VioscreenAdminAPITask_singleton = VioscreenAdminAPITask()
    ... do stuff ...
```

File: microsetta_private_api/util/vioscreen.py (hunk 2 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:166]: trial and error

File: microsetta_private_api/util/vioscreen.py (hunk 3 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:239]: Not with the vioscreen API, at least that im aware of

File: microsetta_private_api/util/vioscreen.py [hunk 11]
```diff
@@ -79,3 +86,221 @@ def decode_key(encoded):
...
+    def dietaryscore(self, id_):
+        return self._get_session_data(id_, 'dietaryscore').get('dietaryScore')
+
+    def users(self):
+        return self.get('users')['users']
+
+    def sessions(self, vioscreen_user):
+        detail = self.get('users/%s/sessions' % vioscreen_user)
+
+        # TODO: Can we push waiting for the async result out to flask to
```
Comment (wasade) [inline microsetta_private_api/util/vioscreen.py:263]: I think we should just have the client poll. I'm much more concerned about blocked webserver processes than the load from polling

File: microsetta_private_api/util/vioscreen.py (hunk 1 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py]: Agree...  I was basically using the continue as a one time GOTO which is always super clunky.  I've moved out the request handling to a separate inner function and split out the update_headers as you indicate, which does clean up the logic a little bit.  I was also thinking I could mess with the retry delay based on which error code came back from vioscreen, so that auth errors retry without delay, but if this mechanism works well enough, I don't see a reason to atm.

File: microsetta_private_api/util/vioscreen.py [hunk 12]
```diff
@@ -79,3 +86,221 @@ def decode_key(encoded):
...
+            data = f(session_id)
+            if data is None:
+                errors.append("FFQ appears incomplete or not taken")
+                break
+
+            results[name] = data
+
+        return errors, results
+
+    def top_food_report(self, session_id):
```
Comment (wasade) [inline microsetta_private_api/util/vioscreen.py:294]: should we store the result in the database so we only fetch from vioscreen once?

**Comment by wasade** [comment]
Thanks, @dhakim87!! I think this looks super cool. I completely get that it is not ideal on the async side of things, but it doesn't look that bad in my opinion if we tolerate a little bit of polling

File: microsetta_private_api/util/vioscreen.py (hunk 2 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py:166]: Vioscreen application layer error codes.  I don't know that we have a full list of them or if these were just the ones we encountered doing initial development against their apis.

File: microsetta_private_api/util/vioscreen.py (hunk 3 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py:239]: Our functions just match the swagger api of the vioscreen website (https://api.viocare.com/ - can send creds or show you their api out of band).  I don't see any reason to diverge from their api here, and I don't know that they have a bulk download url implemented anyway.

File: microsetta_private_api/api/_survey.py (hunk 7 above)
Comment (dhakim87) [reply microsetta_private_api/api/_survey.py:255]: If we want to show the user their pdf in the browser, the current implementation is correct.  If we want the user to download the pdf directly to their computer, the second option would be more correct (though it hardly matters since this gets routed through microsetta-interface anyway, the response headers set in that project are the ones that actually determine what happens)

File: microsetta_private_api/repo/survey_answers_repo.py (hunk 8 above)
Comment (dhakim87) [reply microsetta_private_api/repo/survey_answers_repo.py:47]: Yes please!  I consider this the main reason to hold back this PR: none of our routes are capable of saying that a survey is both a primary survey and a vioscreen survey, but that occurs in the dev db.  If that's also happening on staging/production, we either need to change our model to add a legacy combined survey type, or we need a schema change to split into two identifiers.

File: microsetta_private_api/repo/survey_answers_repo.py (hunk 9 above)
Comment (dhakim87) [reply microsetta_private_api/repo/survey_answers_repo.py:61]: iirc, it's difficult or impossible to determine what survey was taken if the user sends back an empty response in our current database models.  I shifted code around to make surveys that are both primary and vioscreen return as vioscreen rather than primary for testing purposes, and probably mucked up the empty survey handling in the process.  But I want to remove all of that and have us externally resolve the surveys that are both primary and vioscreen because I don't know what to do with the logic here.

File: microsetta_private_api/util/vioscreen.py (hunk 10 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py:112]: afaict, flask instantiates the VioscreenAdminAPITask when make_vioscreen_request is called for the first time.  It should be totally unnecessary to do that, since it's unclear to me how that object would be synchronized with the one in the celery process, but that appears to be the library's behavior.  I don't really care all too much about stopping it from having that behavior- I just can't let flask's workers run off and make requests to vioscreen, so I had to remove the update_headers call from the constructor.  Frankly I think making web requests within the constructor has the potential for badness anyway.  

I think the real issue to address here is on what events should we be refreshing vioscreen credentials on the celery side.  Right now we do it whenever someone makes a request and our auth isn't set or fails, but we could theoretically have a timer that triggers a refresh a few minutes prior to our tokens expiring.  That would give a much better experience to the first poor user who makes vioscreen requests after the token expires, since they wouldn't be waiting for the authentication to go through

File: microsetta_private_api/util/vioscreen.py (hunk 11 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py:263]: I think polling feels like a trap.  It trades cpu, network usage, and increased complexity on the javascript, microsetta-interface, and private-api all because we're trying to mix two libraries with incompatible approaches.  Suddenly embed tags and pdf downloads don't behave in a standard way and the client browser must have javascript enabled.  It adds extra complexity in the client pages and extra complexity in either redis or postgres to store which account made which celery request so end users can wait on them.  

The celery api's are set up for async but flask doesn't have an event loop to make use of it.  I'd prefer we either bite the bullet and go full async, or accept flask's limitations and block.  It's provable that blocking here can't cause deadlocks since celery can't start new flask requests.  So the only issue we run into is if we exhaust the pool of flask workers.  But we have control over how many flask workers are started, and can even configure it to start workers per request if we're worried.  It's a waste of memory allocating all those empty stacks for new threads, but it seems to be the way flask is designed.  

If these requests took minutes or hours or days I could totally understand having the user come back later, but for a pdf download that takes half a second (assuming we manage the vioscreen update_headers calls in a smart way), I just don't see a reason for all that.

File: microsetta_private_api/util/vioscreen.py (hunk 12 above)
Comment (dhakim87) [reply microsetta_private_api/util/vioscreen.py:294]: I'm opposed to polling approaches, but I recognize it would be necessary to cache responses if we were to move to one.  The good part is that we lower the number of requests we make to vioscreen.  The bad part is that I have no idea if vioscreen updates what comes back in the pdf over time, and have no idea what policy should exist to determine when to pull new data from vioscreen and when to use cached data from the database.  I don't believe they have any kind of ETag to check, but I admit I wasn't looking for one.  

If we plan to go back to them and ask them for a way to hide all the N/A and Age information at the top of the pdf reports (that appears since we don't pass them the information they would need to generate this information), caching results locally will complicate updating to the most recent version of their api.

File: microsetta_private_api/api/_survey.py (hunk 7 above)
Comment (wasade) [reply microsetta_private_api/api/_survey.py:255]: Okay, thanks. The download sound more correct from -private-api and if needed the headers are modified in -interface

File: microsetta_private_api/repo/survey_answers_repo.py (hunk 8 above)
Comment (wasade) [reply microsetta_private_api/repo/survey_answers_repo.py:47]: Can you send a query that shows this? I'll investigate a patch

File: microsetta_private_api/util/vioscreen.py (hunk 10 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:112]: That sounds like a good solution. It looks like celery supports periodic tasks natively

File: microsetta_private_api/util/vioscreen.py (hunk 2 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:166]: Just waht was encountered

File: microsetta_private_api/util/vioscreen.py (hunk 11 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:263]: Okay. Let's block and over allocate worker processes.

File: microsetta_private_api/util/vioscreen.py (hunk 12 above)
Comment (wasade) [reply microsetta_private_api/util/vioscreen.py:294]: Okay, let's not cache right now and revisit if needed

File: microsetta_private_api/repo/survey_answers_repo.py (hunk 8 above)
Comment (dhakim87) [reply microsetta_private_api/repo/survey_answers_repo.py:47]: select * from ag_login_surveys left join survey_answers using (survey_id) where vioscreen_status=3 limit 10;

**Comment by wasade** [comment]
Note failures look real here
